firebase-web/src/main/java/io/spine/web/firebase/FirebaseRestClient.java

```diff
@@ -0,0 +1,128 @@
+ * </a>.
+ */
+class FirebaseRestClient implements FirebaseClient {
+
+    /**
+     * The format by which Firebase database nodes are accessible via REST API.
+     *
+     * <p>The first placeholder is the database URL and the second one is a node path.
+     */
+    private static final String NODE_URL_FORMAT = "%s/%s.json";
+
+    /**
+     * The representation of the database {@code null} entry.
+     *
+     * <p>In Firebase the {@code null} node is deemed nonexistent.
+     */
+    @SuppressWarnings("DuplicateStringLiteralInspection")
+    @VisibleForTesting
+    static final String NULL_ENTRY = "null";
+
+    private final String databaseUrl;
+    private final HttpRequestExecutor requestExecutor;
+
+    @VisibleForTesting
+    FirebaseRestClient(String databaseUrl, HttpRequestExecutor requestExecutor) {
+        this.databaseUrl = databaseUrl;
+        this.requestExecutor = requestExecutor;
+    }
+
+    /**
+     * Creates a {@code FirebaseRestClient} which operates on the database located at given
+     * {@code databaseUrl} and uses given {@code httpTransport}.
+     */
+    static FirebaseRestClient create(String databaseUrl, HttpTransport httpTransport) {
+        HttpRequestExecutor requestExecutor = HttpRequestExecutor.using(httpTransport);
+        return new FirebaseRestClient(databaseUrl, requestExecutor);
+    }
+
+    @Override
+    public Optional<FirebaseNodeValue> get(FirebaseDatabasePath nodePath) {
+        checkNotNull(nodePath);
+
+        GenericUrl url = toNodeUrl(nodePath);
+        String data = requestExecutor.get(url);
+        if (isNullData(data)) {
+            return Optional.empty();
+        }
+        FirebaseNodeValue value = FirebaseNodeValue.from(data);
+        Optional<FirebaseNodeValue> result = Optional.of(value);
+        return result;
+    }
+
+    @Override
+    public void addValue(FirebaseDatabasePath nodePath, FirebaseNodeValue value) {
+        checkNotNull(nodePath);
+        checkNotNull(value);
+
+        GenericUrl url = toNodeUrl(nodePath);
+        ByteArrayContent byteArrayContent = value.toByteArray();
+        Optional<FirebaseNodeValue> existingValue = get(nodePath);
+        if (!existingValue.isPresent()) {
+            create(url, byteArrayContent);
+        } else {
+            update(url, byteArrayContent);
+        }
+    }
+
+    /**
+     * Creates the database node with the given value or overwrites the existing one.
+     */
+    private void create(GenericUrl nodeUrl, HttpContent value) {
+        requestExecutor.put(nodeUrl, value);
+    }
+
+    /**
+     * Updates the database node with the given value.
+     *
+     * <p>Common entries are overwritten.
+     */
+    private void update(GenericUrl nodeUrl, HttpContent value) {
+        requestExecutor.patch(nodeUrl, value);
+    }
+
+    private GenericUrl toNodeUrl(FirebaseDatabasePath nodePath) {
+        String url = String.format(NODE_URL_FORMAT, databaseUrl, nodePath);
```

Comment:
It seems to me that the performance could be better here. `databaseUrl` is the `final` field, and you can format it once the instance is created. So the `nodePath` is the only value to use in formatting.